torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -718,6 +733,132 @@ def call_function(
                 return TupleVariable([TupleVariable(items), aux])
 
 
+class FunctorchVmapHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    def call_function(
+        self, tx, args: "List[VariableTracker]", kwargs: "Dict[str, VariableTracker]"
+    ) -> "VariableTracker":
+        from . import ConstantVariable, TensorVariable
+        from .builder import wrap_fx_proxy
+
+        if not torch._dynamo.config.capture_func_transforms:
+            unimplemented("torch.func.vmap capture is disabled")
+
+        checkpoint = tx.copy_graphstate()
+        graph_checkpoint = tx.output.graph
+
+        # unpack args
+        fn = args[0]
+        in_dims = args[1]
+        out_dims = args[2]
+        randomness = args[3]
+        chunk_size = args[4]
+        batch_input_args = args[5:]
+
+        if kwargs:
+            unimplemented(
+                "NYI - torch.func.vmap: kwargs arguments are currently unsupported."
+            )
+
+        if chunk_size.value is not None:
+            unimplemented(
+                "NYI - torch.func.vmap is not implemented when chunk_size is passed"
+            )
+
+        flat_args, arg_spec = torch.utils._pytree.tree_flatten(batch_input_args)
+        in_dims_v = in_dims.as_python_constant()
+        in_dims_v = in_dims_v if isinstance(in_dims_v, int) else list(in_dims_v)
+        broadcasted_in_dims = torch._functorch.vmap._broadcast_to_and_flatten(
+            in_dims_v, arg_spec
+        )
+
+        # We want to pass unbatched input to speculate subgraph.
+        # So we loop through the inputs and select only one sample
+        # from the batch.
+        unbatched_input_args = []
+        for arg, in_dim in zip(flat_args, broadcasted_in_dims):
+            if in_dim is not None:
+                assert isinstance(arg, TensorVariable)
+                unbatched_arg = arg.call_method(
+                    tx, "select", (ConstantVariable(in_dim), ConstantVariable(0)), {}
+                )
+                unbatched_input_args.append(unbatched_arg)
+            else:
+                unbatched_input_args.append(arg)
+
+        # trace through the function with unbatched inputs.
+        _, body_graph, body_lifted_freevars = speculate_subgraph(
+            tx,
+            fn,
+            torch.utils._pytree.tree_unflatten(unbatched_input_args, arg_spec),
+            {},
+            graph_checkpoint,
+            checkpoint,
+        )
```

Comment:
We need to make sure .stride() and .storage_offset() (are there more?) are never accessed in this `speculate_subgraph`. Otherwise, the captured graph might be wrong, because vmap has different behavior w.r.t.  strides and storage offset than regular eager-mode pytorch.